Fix read describe cases (#39)
epiverse-trace/tutorials-early: episodes/read-cases.Rmd

## Reading from files 

Several packages are available for importing outbreak data stored in individual files into `R`. These include [rio](http://gesistsa.github.io/rio/), [readr](https://readr.tidyverse.org/) from the `tidyverse`, [io](https://bitbucket.org/djhshih/io/src/master/), [ImportExport](https://cran.r-project.org/web/packages/ImportExport/index.html), [data.table](https://rdatatable.gitlab.io/data.table/). Together, these packages offer methods to read single or multiple files in a wide range of formats.

The below example shows how to import a `csv` file into `R` environment using `{rio}` package.

```{r,warning=FALSE,message=FALSE}
library("rio")
library("here")
# read data

> case_data <- rio::import(here::here("built", "data", "ebola_cases.csv"))

Author: Good point. Thanks.

::::::::::::::::::::::::::::::::: challenge

###  Reading compressed data 

Take 1 minute:
- Is it possible to read compressed data in `R`?

::::::::::::::::: hint

> You can check the supported file formats in the `{rio}` package as follows:

Author: Thank you.

::::::::::::::::: solution

```{r,eval=FALSE}

> rio::import(here::here("some", "where", "downto", "path", "file_name.zip"))

Author: Ok. I have a Marburg outbreak data set to the data.

The [DBI](https://dbi.r-dbi.org/) package serves as a versatile interface for interacting with database management systems (DBMS) across different back-ends or servers. It offers a uniform method for accessing and retrieving data from various database systems.

The following code chunk demonstrates how to create a temporary SQLite database in memory, store the `case_data` as a table within it, and subsequently read from it:

```{r,warning=FALSE,message=FALSE}

> requireNamespace("DBI", quietly = TRUE)
> requireNamespace("RSQLite", quietly = TRUE)

Reviewer: as suggested above

```suggestion
library(DBI)
library(RSQLite)

```

> # Create a temporary SQLite database in memory
> db_con <- DBI::dbConnect(RSQLite::SQLite(), ":memory:")
> # Store the 'case_data' dataframe as a table named 'cases'

Reviewer: In this chuck, adding a space between each action increases the visibility of each step. We can replicate this in the following lines too (read, close, view)

```suggestion
# Create a temporary SQLite database in memory
db_con <- DBI::dbConnect(RSQLite::SQLite(), ":memory:")

# Store the 'case_data' dataframe as a table named 'cases'
```

The following code chunk demonstrates how to create a temporary SQLite database in memory, store the `case_data` dataframe as a table within it, and subsequently read from it:

```{r,warning=FALSE,message=FALSE}

> requireNamespace("DBI", quietly = TRUE)

Reviewer 2: Is this section not covered by {readepi}?

{readepi} can read from RDBMS and APIs

Reviewer: I'm interested to read the rationale behind this decision too

Author: Yes, {readepi} can import data from RDBMS, however,  by the time I wrote this episode this functionality was not full implemented. Later when it is full functional we can update this section. Also, {DBI} allows for creating light in-memory database, which fit our purpose her.

base::print(utils::head(result))
```

> This code first establishes a connection to an SQLite database created in memory using `dbConnect()`. Then, it writes the `case_data` into a table named 'cases' within the database using the `dbWriteTable()` function. Subsequently, it reads the data from the 'cases' table using `dbReadTable()`. Finally, it closes the database connection with `dbDisconnect()`. More examples about SQL databases and R can be found [here](https://datacarpentry.org/R-ecology-lesson/05-r-and-databases.html).

Reviewer: After learning how to connect with DB's it is useful to learn the rationale behind using SQL queries to optimize local memory usage. I suggest to add a callout and redirect to the {dbplyr} package, which is also used in the datacarpentries episode already linked there.

```suggestion
This code first establishes a connection to an SQLite database created in memory using `dbConnect()`. Then, it writes the `case_data` into a table named 'cases' within the database using the `dbWriteTable()` function. Subsequently, it reads the data from the 'cases' table using `dbReadTable()`. Finally, it closes the database connection with `dbDisconnect()`. More examples about SQL databases and R can be found [here](https://datacarpentry.org/R-ecology-lesson/05-r-and-databases.html).

:::::::::::::::::::::: callout

### Run SQL queries in R using dbplyr

Using a database interface package is invaluable for optimizing memory usage because it can strategically limit the amount of data pulled from the database through queries. By leveraging such a package, one can ensure that most aggregation tasks are performed directly within the database before extracting the necessary information. This approach minimizes the need to load large datasets into local memory for processing, thereby conserving resources and enhancing performance. Conversely, relying solely on external data manipulation processes can lead to suboptimal memory usage. Conducting all data operations outside the database could tax system resources and hinder efficiency.

Read the [Introduction to dbplyr](https://dbplyr.tidyverse.org/articles/dbplyr.html) vignette to learn how to generate your own queries!

::::::::::::::::::::::

```

Author: Thank you for your suggestion.

## Reading from HIS APIs

Health related data are also increasingly stored in specialized HIS APIs like **Fingertips**, **GoData**, **REDCap**, and **DHIS2**. In such case one can resort to [readepi](https://epiverse-trace.github.io/readepi/) package, which enables reading  data from HIS-APIs.  

> -[TBC]

Reviewer: Is this going to be covered in this PR or in a separate one?

Author: This will be covered by a separate PR.

base::print(utils::head(result))
```

This code first establishes a connection to an SQLite database created in memory using `dbConnect()`. Then, it writes the `case_data` into a table named 'cases' within the database using the `dbWriteTable()` function. Subsequently, it reads the data from the 'cases' table using `dbReadTable()`. Finally, it closes the database connection with `dbDisconnect()`. More examples about SQL databases and R can be found [here](https://datacarpentry.org/R-ecology-lesson/05-r-and-databases.html).

:::::::::::::::::::::: callout

### Run SQL queries in R using dbplyr

> Using a database interface package is invaluable for optimizing memory usage because it can strategically limit the amount of data pulled from the database through queries. By leveraging such a package, one can ensure that most aggregation tasks are performed directly within the database before extracting the necessary information. This approach minimizes the need to load large datasets into local memory for processing, thereby conserving resources and enhancing performance. Conversely, relying solely on external data manipulation processes can lead to suboptimal memory usage. Conducting all data operations outside the database could tax system resources and hinder efficiency.

Reviewer: I think I added this and looks too long. Two sentences may be enough for this.

```suggestion
A database interface package optimize memory usage by processing the database before extraction, reducing memory load. Conversely, conducting all data manipulation outside the database (e.g., in our local Rstudio session) can lead to inefficient memory usage and strained system resources.
```
